### PR TITLE
Move some LValue syntax checks out of hte parser.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-compiler @jcouv @AlekseyTs

**Comment by CyrusNajmabadi** [comment]
@gafter @cston can you take a look?

File: src/Compilers/CSharp/Test/Syntax/Parsing/DeconstructionTests.cs [hunk 1]
```diff
@@ -2144,42 +2144,232 @@ void M()
...
+        [Fact]
+        public void NoDeconstructionAsLvalue_7()
+        {
+            var source =
+@"
+class C
+{
+    void M(string e)
+    {
         ref int x = ref var(x, y); // error 7
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Syntax/Parsing/DeconstructionTests.cs]: > // error 7 [](start = 34, length = 11)

Consider removing the ` // error <number>` comments now the test is split up.

**Comment by CyrusNajmabadi** [comment]
@ManishJayaswal 

Customer scenario

None. This is general infrastructure work for the compiler based on design changes the team agreed on.

Bugs this fixes:

N/A

Workarounds, if any

N/A

Risk

Low.

Performance impact

Low.

Is this a regression from a previous update?

N/A

Root cause analysis:

N/A. 

How was the bug found?

N/A

**Comment by CyrusNajmabadi** [comment]
Rebasing to 15.6

**Comment by CyrusNajmabadi** [comment]
The CSharpNavigateTo.NavigateTo failure is tracked with https://github.com/dotnet/roslyn/issues/19530

**Pull request merged**